File: misc/plugin/preview.rb
# -*- coding: utf-8; -*-
#
# preview.rb: view preview automatically
#
# Copyright (c) MATSUOKA Kohei <http://www.machu.jp/>
# Distributed under the GPL2 or any later version.
#

@conf['preview.interval'] ||= 10
@conf['preview.min_width'] ||= 896

if /\A(form|edit|preview)\z/ === @mode then
	enable_js('preview.js')
	add_js_setting('$tDiary.plugin.preview')
	add_js_setting('$tDiary.plugin.preview.interval', @conf['preview.interval'].to_json)
	add_js_setting('$tDiary.plugin.preview.minWidth', @conf['preview.min_width'].to_json)
end

add_conf_proc('preview', @preview_label_conf, 'update') do
	if @mode == 'saveconf'
		@conf['preview.interval'] = @cgi.params['preview.interval'][0].to_i
		@conf['preview.min_width'] = @cgi.params['preview.min_width'][0].to_i
	end
	ERB.new(%q{
		<h3><%= @preview_label_interval %></h3>
		<div>
			<p><%= @preview_label_interval_desc %></p>
			<p><input name="preview.interval" value="<%=h @conf['preview.interval'] %>"> sec</p>
		</div>
		<h3><%= @preview_label_min_width %></h3>
		<div>
			<p><%= @preview_label_min_width_desc %></p>
			<p><input name="preview.min_width" value="<%=h @conf['preview.min_width'] %>"> px</p>
		</div>
	}).result(binding)
end

File: js/preview.js
/*
 * preview.js: view preview automatically
 *
 * Copyright (c) MATSUOKA Kohei <http://www.machu.jp/>
 * Distributed under the GPL2 or any later version.
 */
$(function() {

var previewButton = $('input[name*="preview"]');
var intervalRate = 1;

$tDiary.plugin.preview.reload = function() {
  previewButton.prop("disabled", true);
  $.post(
    'update.rb',
    $('form.update').serialize() + "&appendpreview=1",
    function(data) {
      var previewContent = $(data).find('div.autopagerize_page_element');
      if (previewContent.size() != 0) {
        $('div.autopagerize_page_element').replaceWith(previewContent);
        intervalRate = 1;
      } else {
        intervalRate *= 2;
        console.info('[preview.js] update failed, the next update will be after '
          + $tDiary.plugin.preview.interval * intervalRate + 'sec.');
      }
      $('div.day')
        .css('flex', "1 1 " + $tDiary.plugin.preview.minWidth / 2 + "px");
    },
    'html'
  )
  .fail(function() {
    intervalRate *= 2;
    console.info('[preview.js] update failed, the next update will be after '
      + $tDiary.plugin.preview.interval * intervalRate + 'sec.');
  })
  .always(function() {
    previewButton.prop("disabled", false);
    setTimeout($tDiary.plugin.preview.reload,
      $tDiary.plugin.preview.interval * 1000 * intervalRate);
  });
}

if ($('div.autopagerize_page_element').length == 0) {
  $('div.update').before(
    '<div class="day autopagerize_page_element">'
  );
}

$('<div class="preview-container"></div>')
  .css('display', 'flex')
  .css('flex-flow', 'row-reverse wrap')
  .insertAfter('h1')
  .append($('div.day'));
$('div.day')
  .css('flex', "1 1 " + $tDiary.plugin.preview.minWidth / 2 + "px");
$("div.whole-content").css('max-width', 'none');

// プレビューボタンを押した時もajaxで更新するよう設定
previewButton.click(
  function(event) {
    event.preventDefault();
    $tDiary.plugin.preview.reload();
  }
);

$tDiary.plugin.preview.reload();

});
